test/lsp.py

```diff
@@ -8,13 +8,76 @@
 import sys
 import traceback
 import re
-
+import tty
+import functools
+from collections import namedtuple
+import queue
+import threading
+from copy import deepcopy
 from typing import Any, List, Optional, Tuple, Union
 from enum import Enum, auto
+from itertools import islice
 
 import colorama # Enables the use of SGR & CUP terminal VT sequences on Windows.
 from deepdiff import DeepDiff
 
+TestRegexesTuple = namedtuple("TestRegexesTuple", [
+    "sendRequest",
+    "findQuotedTag",
+    "findTag",
+    "fileDiagnostics",
+    "diagnostic"
+])
+TEST_REGEXES = TestRegexesTuple(
+    re.compile(R'^// -> (?P<method>[\w\/]+) {'),
+    re.compile(R'(?P<tag>"@\w+")'),
+    re.compile(R'(?P<tag>@\w+)'),
+    re.compile(R'// (?P<testname>\w+):[ ]?(?P<diagnostics>[\w @]*)'),
+    re.compile(R'(?P<tag>@\w+) (?P<code>\d\d\d\d)')
+)
+
+TagRegexesTuple = namedtuple("TagRegexestuple", ["simpleRange", "multilineRange"])
+TAG_REGEXES = TagRegexesTuple(
+    re.compile(R"(?P<range>[\^]+) (?P<tag>@\w+)"),
+    re.compile(R"\^(?P<delimiter>[()]{1,2}) (?P<tag>@\w+)$")
+)
+
+
+def countIndex(sequence, start=0):
+    n = start
+    for elem in sequence:
+        yield n, elem
+        n += 1 + len(elem)
+
+# Filter the lines for tag comments and report line number that tags refer to
+def tagsOnly(lines, start=0):
+    n = start
+    numCommentLines = 0
+
+    def hasTag(line):
+        if line.find("// ") != -1:
+            for kind, regex in TAG_REGEXES._asdict().items():
+                if regex.search(line[len("// "):]) != None:
+                    return True
+        return False
+
+    for line in lines:
+        if hasTag(line):
+            numCommentLines += 1
+            yield n - numCommentLines, line
+        else:
+            numCommentLines = 0
+
+        n += 1
+
+
+def preceedComments(sequence):
```

Comment:
Just scrolling through the code. This is really cool PR! What is `preceed`? Is that a typo?

Reply:
No, that's deliberate: https://www.thefreedictionary.com/preceed "to put in front" "to put before" basically :)

Reply:
Hehe :) Ah! Looks like that there are different (correct) ways to write that. I thought it is written `precede`.

Reply:
God damn, you're right. I was so sure.

Reply:
if that's so complex, amybe next time just use what is well used already (prepend vs append) :)